# Imports

In [7]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn import metrics as mt

# Load dataset

In [2]:
df = pd.read_csv("../dataset/train.csv")

# Data preparation

In [3]:
df.head()

id_cliente  idade  saldo_atual  divida_atual  renda_anual  \
0        1767     21   278.172008       2577.05  24196.89636   
1       11920     40   268.874152       2465.39  19227.37796   
2        8910     36   446.643127       1055.29  42822.28223   
3        4964     58   321.141267        703.05  51786.82600   
4       10100     35   428.716114        891.29  44626.85346   

   valor_em_investimentos  taxa_utilizacao_credito  num_emprestimos  \
0              104.306544                31.038763                6   
1               69.858778                36.917093                5   
2              134.201478                34.561714                0   
3              297.350067                31.493561                0   
4              134.201478                28.028887                2   

   num_contas_bancarias  num_cartoes_credito  dias_atraso_dt_venc  \
0                     5                    7                   21   
1                     8                    5                   40   
2                     3                    6                   26   
3                     3                    7                   12   
4                     8                    7                   24   

   num_pgtos_atrasados  num_consultas_credito  taxa_juros investe_exterior  \
0                   14                      9          15              Não   
1                   23                     10          18              Não   
2                   13                      3          15              Sim   
3                    7                      2           1              Sim   
4                   10                      8          20              Sim   

  pessoa_polit_exp limite_adicional  
0              Não            Negar  
1              Não            Negar  
2              Não            Negar  
3              Não            Negar  
4              Não            Negar

In [4]:
df.columns

Index(['id_cliente', 'idade', 'saldo_atual', 'divida_atual', 'renda_anual',
       'valor_em_investimentos', 'taxa_utilizacao_credito', 'num_emprestimos',
       'num_contas_bancarias', 'num_cartoes_credito', 'dias_atraso_dt_venc',
       'num_pgtos_atrasados', 'num_consultas_credito', 'taxa_juros',
       'investe_exterior', 'pessoa_polit_exp', 'limite_adicional'],
      dtype='object')

In [3]:
features = ['idade', 'divida_atual', 'renda_anual',
            'valor_em_investimentos', 'taxa_utilizacao_credito', 'num_emprestimos',
            'num_contas_bancarias', 'num_cartoes_credito', 'dias_atraso_dt_venc',
            'num_pgtos_atrasados', 'num_consultas_credito', 'taxa_juros']

label = ['saldo_atual']

In [4]:
x_train = df.loc[:, features]
y_train = df.loc[:, label]

In [7]:
x_train

idade  divida_atual   renda_anual  valor_em_investimentos  \
0        21       2577.05   24196.89636              104.306544   
1        40       2465.39   19227.37796               69.858778   
2        36       1055.29   42822.28223              134.201478   
3        58        703.05   51786.82600              297.350067   
4        35        891.29   44626.85346              134.201478   
...     ...           ...           ...                     ...   
9495     29        157.98   32624.67754              237.739275   
9496   1237        805.43   49024.15700              118.883577   
9497     47       2250.42   42200.88978              134.201478   
9498     42        505.37   35391.32289              245.498881   
9499     35        851.22  106192.53216               78.140488   

      taxa_utilizacao_credito  num_emprestimos  num_contas_bancarias  \
0                   31.038763                6                     5   
1                   36.917093                5                     8   
2                   34.561714                0                     3   
3                   31.493561                0                     3   
4                   28.028887                2                     8   
...                       ...              ...                   ...   
9495                32.407322              100                     4   
9496                23.632195                0                     4   
9497                34.623576                4                     4   
9498                28.758818                3                     7   
9499                30.528134                0                     0   

      num_cartoes_credito  dias_atraso_dt_venc  num_pgtos_atrasados  \
0                       7                   21                   14   
1                       5                   40                   23   
2                       6                   26                   13   
3                       7                   12                    7   
4                       7                   24                   10   
...                   ...                  ...                  ...   
9495                    4                   24                   15   
9496                    7                   10                   11   
9497                    4                   13                    8   
9498                    6                   24                   13   
9499                    6                    8                    4   

      num_consultas_credito  taxa_juros  
0                         9          15  
1                        10          18  
2                         3          15  
3                         2           1  
4                         8          20  
...                     ...         ...  
9495                      1          14  
9496                      2           3  
9497                      4          24  
9498                      3          11  
9499                   2377           3  

[9500 rows x 12 columns]

In [8]:
y_train

saldo_atual
0      278.172008
1      268.874152
2      446.643127
3      321.141267
4      428.716114
...           ...
9495   157.500279
9496   497.714090
9497   306.557684
9498   209.870718
9499   971.407512

[9500 rows x 1 columns]

# Model Training

In [5]:
# model definition
lr_model = LinearRegression()

# model fit (ajuste, treinamento)
lr_model.fit(x_train, y_train)

#previsao
y_pred = lr_model.predict(x_train)

In [6]:
df1 = df.loc[:, ['id_cliente', 'saldo_atual']]
df1['predicted'] = y_pred

df1.head()

id_cliente  saldo_atual   predicted
0        1767   278.172008  346.669549
1       11920   268.874152  367.840277
2        8910   446.643127  431.468979
3        4964   321.141267  445.506463
4       10100   428.716114  378.271169

In [15]:
lr_model.coef_

array([[ 3.96706202e-04, -4.00595601e-02,  2.77622532e-06,
        -1.04318668e-03,  9.80890872e+00, -1.22353405e-02,
        -6.33015538e-03, -3.57808095e-03, -2.15858165e+00,
         3.77570060e-04,  6.79176336e-03,  3.48471845e-03]])

In [17]:
lr_model.intercept_

array([190.82359144])

In [18]:

np.sum( x_train.loc[0, ].values * lr_model.coef_ ) + lr_model.intercept_

array([346.66954862])

In [ ]:
y = lr_model.intercept_ + lr_model.coef_[0][0] * x_train.iloc[0, 0] + a2 * x2 + ...

coeficientes da reta de regressão linear
 y = a + bx
y = a0 + a1x1 + a2x2 + a3x3 + ... + a12x12

a0 = intercepto da reta = 190.82
a1 = coeficiente de idade
a2 = coeficiente de divida_atual
...
a12 = coeficiente de taxa_juros

x1 = idade
x2 = divida_atual
x3 = renda_anual
x4 = valor_em_investimentos
...
x12 = taxa_juros

# Modal Performance

## R squared

In [10]:
r2_score = mt.r2_score(y_train, y_pred)
print(f'R2: {r2_score}')
print( '{}% da variação da variável alvo y é reduzida, levando em consideração o preditor'.format( r2_score ) )
print( '{}% da variação da variável alvo y é “explicada pela variação do preditor x'.format( r2_score ) )

R2: 0.16917364489050024
0.16917364489050024% da variação da variável alvo y é reduzida, levando em consideração o preditor
0.16917364489050024% da variação da variável alvo y é “explicada pela variação do preditor x


## MSE - Mean Squared Error 
 - média do quadrado das difernças entre os valores preditos e os valores reais

In [11]:
mse = mt.mean_squared_error(y_train, y_pred)
print(f'A cada previsão, o erro médio quadrático é de {mse}')

A cada previsão, o erro médio quadrático é de 39370.26503401534


# RMSE - Root Mean Squared Error
 - raiz quadrada da média do quadrado das difernças entre os valores preditos e os valores reais

In [12]:
rmse = np.sqrt(mse)
print(f'A cada previsão, o erro médio é de: U${rmse}')

A cada previsão, o erro médio é de: U$198.4194169783173
